server/s3_data_usage.cc

```diff
@@ -0,0 +1,511 @@
+           "Data usage item for key=%s will be removed from cache as failed",
+           item->cache_key.c_str());
+    items.erase(item->cache_key);
+  }
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+}
+
+void S3DataUsageCache::update_data_usage(std::shared_ptr<RequestObject> req,
+                                         std::shared_ptr<S3BucketMetadata> src,
+                                         int64_t objects_count_increment,
+                                         int64_t bytes_count_increment,
+                                         std::function<void()> on_success,
+                                         std::function<void()> on_failure) {
+  const std::string &req_id = req->get_stripped_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry", __func__);
+  std::shared_ptr<DataUsageItem> item(get_item(req));
+  if (item) {
+    item->save(req, objects_count_increment, bytes_count_increment, on_success,
+               on_failure);
+  } else {
+    s3_log(S3_LOG_DEBUG, req_id, "Failed to retrieve a Data usage cache item");
+    on_failure();
+  }
+
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit", __func__);
+}
+
+DataUsageItem::DataUsageItem(std::shared_ptr<RequestObject> req,
+                             const std::string &key_in_cache,
+                             DataUsageStateNotifyCb subscriber) {
+  s3_log(S3_LOG_DEBUG, "", "%s Entry\n", __func__);
+  current_request = std::move(req);
+  pending_request = nullptr;
+  cache_key = std::move(key_in_cache);
+  motr_key = cache_key + "/" + get_server_id();
+  state_notify = subscriber;
+  state = DataUsageItemState::empty;
+  current_objects_increment = 0;
+  current_bytes_increment = 0;
+  pending_objects_increment = 0;
+  pending_bytes_increment = 0;
+  s3_log(S3_LOG_DEBUG, "", "%s Exit\n", __func__);
+}
+
+std::string DataUsageItem::get_item_request_id() {
+  return current_request ? current_request->get_stripped_request_id() : "";
+}
+
+void DataUsageItem::set_state(DataUsageItemState new_state) {
+  std::string req_id = get_item_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+  if (new_state != state) {
+    DataUsageItemState old_state = state;
+    state = new_state;
+    state_notify(this, old_state, new_state);
+  }
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+}
+
+void DataUsageItem::save(std::shared_ptr<RequestObject> req,
+                         int64_t objects_count_increment,
+                         int64_t bytes_count_increment,
+                         std::function<void(void)> on_success,
+                         std::function<void(void)> on_failure) {
+  const std::string &req_id = req->get_stripped_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+
+  // object_list_index_layout should be set before using this method
+  // assert(non_zero(data_usage_accounts_index_layout.oid));
+  s3_log(S3_LOG_INFO, req_id,
+         "%s Object count increment %ld; bytes count increment %ld", __func__,
```

Comment:
`%ld` is wrong, see proper macro for `int64_t`, should be something like `PRId64`.

Reply:
Done.